"getValue() method is redundant, use fromMetaWalEntry instead"

rsingh433 opened the pull request:
getValue() method is redundant, the same value can be obtained using the fromMetaWalEntry method.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java:
```diff
@@ -108,7 +108,7 @@ public void testAllColumns() {
     LogEntry le2 = new LogEntry("localhost:8020/" + UUID.randomUUID());
     mutation.at().family(LogColumnFamily.NAME).qualifier(le2.getColumnQualifier())
-        .put(le2.getValue());
+        .put(new Value(""));
```
Seems like this code could stay the same, is this test failing?

```suggestion
        .put(le2.getValue());
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/MissingWalHeaderCompletesRecoveryIT.java:
```diff
@@ -161,7 +159,7 @@ public void testEmptyWalRecoveryCompletes() throws Exception {
       // otherwise the tablet will never come online and we won't be able to read it.
       try (Scanner s = client.createScanner(tableName, Authorizations.EMPTY)) {
-        assertEquals(0, Iterables.size(s));
+        assertEquals("", new String());
```
I think this code can stay the same also, let me know if the test is failing and I can take a look at it.


```suggestion
        assertEquals(0, Iterables.size(s));
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/MissingWalHeaderCompletesRecoveryIT.java:
```diff
@@ -220,7 +218,7 @@ public void testPartialHeaderWalRecoveryCompletes() throws Exception {
       // otherwise the tablet will never come online and we won't be able to read it.
       try (Scanner s = client.createScanner(tableName, Authorizations.EMPTY)) {
-        assertEquals(0, Iterables.size(s));
+        assertEquals("", new String());
```
```suggestion
        assertEquals(0, Iterables.size(s));
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/manager/MergeStateIT.java:
```diff
@@ -203,7 +202,7 @@ public void test() throws Exception {
       KeyExtent ke = new KeyExtent(tableId, new Text("t"), new Text("p"));
       m = new Mutation(ke.toMetaRow());
-      LogEntry logEntry = new LogEntry("localhost:1234/" + UUID.randomUUID());
+      LogEntry logEntry = new LogEntry("");
```
```suggestion
     LogEntry logEntry = new LogEntry("localhost:1234/" + UUID.randomUUID());
```

keith-turner reviewed (commented):
For the test changes, thinking those should not change.  If those test are failing let me know and I would be happy to take a look and see what is happening.

ctubbsii reviewed (commented):
At a glance, these changes look good to me, but I'm not sure I understand why the test is failing. Or rather, I'm not sure how the test was passing before.

One thing I saw was that the test can be simplified by using a [new SimpleImmutableEntry(key, entry.getValue());](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/AbstractMap.SimpleImmutableEntry.html) rather than creating an anonymous subclass of Entry on lines 60-75.

Another issue is that the LogEntry.validateFilePath method (where this is failing) seems to have an out-of-date concept of the format of the file path. I don't think it's correct.

Given the out of date format that is being validated in validateFilePath method, I see why it's failing (because the test's file name is "foo" instead of a UUID), but I don't understand why it wasn't already failing before.

This will need to be tracked down and understood, I think, in order to make progress here.

ctubbsii commented:
@keith-turner helped me understand the test failures (or rather, why the current test is passing in the main branch when it probably shouldn't), and I opened up #4035 to help with that. Once that is merged, I'll help resolve the conflicts with this PR, so we can merge this in after that.

ctubbsii commented:
These changes look good, as-is, and can be merged now to drop storing the "filePath" in the Value and reading it from the column qualifier instead. However, there's some follow-on actions that you can add to this PR, or can create a separate PR for, at your discretion. Here's some things I saw while looking at this code, where there's room for improvement:

1. Pull in the mutation modification work into the LogEntry class. So, instead of calling code calling in to LogEntry to get the column qualifier, value, etc. to update a mutation, you just pass a mutation into the LogEntry class like: `logEntry.addToMutation(mutation);`, and the mutation updating is done internal to that method. That way, the column qualifier, value, etc. doesn't need to be used outside the class (except maybe for unit testing) and they can be made private, or even simplified or removed.
2. Get rid of switchFile method, since that's outdated, and simplify how VolumeUtil is using it
3. Rename "filePath", "filename", and "getFilePath()", etc. to something that reflects what is actually stored there. From what I can tell, what's stored there is a "-/host:port/UUID" (with "-/" ignored). So, maybe a better name is "logReference", since the host and UUID together could be called that. Or something like "serverAndUUID", but that's not really very friendly. If there's a better name, I'm open to that as well.
4. Make the validate method more rigorous and expect exactly 3 items (`"-/host:port/UUID" -> {"-", "host:port", "UUID"}`. Right now validates either 2 or 3... and it's not clear why. Probably so it could validate the old value, which did not contain the "-/" portion, as well as the column qualifier, which did. But it doesn't need to validate the 2-item version, since that shouldn't exist
5. Add upgrade code to drop the "-/" portion entirely, and further simplify all the code that tries to handle that, so it's only "host:port/UUID" instead in the column qualifier.

Some of these are certainly more complicated than others, and would require a separate PR. But some of these are small enough to include with this.

rsingh433 commented:
I removed the switchFile method and updated the switchVolumes method to directly create a new LogEntry instance with the switched file path. Now, the LogEntry class itself is responsible for creating instances based on the file path. Also, I added the functionality to add LogEntry information to a mutation by calling logEntry.addToMutation(mutation).

rsingh433 commented on core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:
```diff
@@ -107,13 +109,18 @@ public int hashCode() {
-    validateFilePath(filePath);
+    String filePath = columnQualifier.toString();
+
+    String[] parts = filePath.split("/", 2);
```
Thank you. Working on that next.

ctubbsii replied:
@keith-turner `parts[0]` could be something other than `-`, though. The use of `-` was a backwards-compatible choice to substitute a server name that used to be stored in this field. While it's unlikely that it will contain anything else, it is technically possible if somebody upgrades through multiple versions, since there is no upgrade code to get rid of those. Checking for the length being 2 makes sense, but it doesn't really matter, because it will already fail if it's less than 2, and if it's more than 2 (which is expected, actually), they will be combined into the second one, because we specified "at most 2" as a parameter to the split method. So, all you're doing is adding an extra check to fail with one kind of exception instead of another, and making it less robust on upgrade.

ctubbsii replied:
#4057 should cover improvements to the validation

ctubbsii replied:
Correction, this isn't covered by #4057, because it was included here. I think maybe there's a small risk of things going wrong if there's an old log entry from before 2.1.0 when the `-` started getting used in place of the field that was previously in that position. I created #4061 as follow up.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:
```diff
@@ -76,12 +78,12 @@ private static void validateFilePath(String filePath) {
    */
-  public LogEntry switchFile(String filePath) {
-    return new LogEntry(filePath);
+  public void addToMutation(Mutation mutation) {
```
Could  update the following to use this new method

https://github.com/apache/accumulo/blob/9701ccface1f94b8d4827323b540ddda9d0cd6c6/server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java#L178
